core/src/Kokkos_Graph.hpp

```diff
@@ -0,0 +1,288 @@
+  GraphBuilder& operator=(GraphBuilder const&) = default;
+  GraphBuilder& operator=(GraphBuilder&&) noexcept = default;
+
+  ~GraphBuilder() = default;
+
+  // </editor-fold> end ctors, destructor, and assignment }}}2
+  //----------------------------------------------------------------------------
+
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="public accessors"> {{{2
+
+  constexpr auto const& get_root() const { return m_root; }
+
+  // </editor-fold> end public accessors }}}2
+  //----------------------------------------------------------------------------
+
+  template <class... PredecessorRefs>
+  // constraints (not intended for subsumption, though...)
+  //   ((remove_cvref_t<PredecessorRefs> is a specialization of
+  //        GraphNodeRef with get_root().get_graph_impl() as its GraphImpl)
+  //      && ...)
+  auto when_all(PredecessorRefs&&... arg_pred_refs) const {
+    // TODO @graph @desul-integration check the constraints and preconditions
+    //                                once we have folded conjunctions from
+    //                                desul
+    auto graph_ptr_impl = get_root().get_graph_ptr();
+    auto node_ptr_impl = graph_ptr_impl->create_aggregate_ptr(arg_pred_refs...);
+    graph_ptr_impl->add_node(node_ptr_impl);
+    KOKKOS_IMPL_FOLD_COMMA_OPERATOR(graph_ptr_impl->add_predecessor(
+        node_ptr_impl, arg_pred_refs) /* ... */);
+    return Kokkos::Impl::GraphAccess::make_graph_node_ref(
+        graph_ptr_impl, std::move(node_ptr_impl));
+  }
+
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="Methods forward to their then_* analogs on root"> {{{2
+
+  template <class... Args>
+  auto parallel_for(Args&&... args) const {
+    return get_root().then_parallel_for((Args &&) args...);
+  }
+  template <class... Args>
+  auto parallel_reduce(Args&&... args) const {
+    return get_root().then_parallel_reduce((Args &&) args...);
+  }
+  template <class... Args>
+  auto parallel_scan(Args&&... args) const {
+    return get_root().then_parallel_scan((Args &&) args...);
+  }
+  template <class... Args>
+  auto deep_copy(Args&&... args) const {
+    return get_root().then_deep_copy((Args &&) args...);
+  }
+
+  // </editor-fold> end Methods forward to their then_* analogs on root }}}2
+  //----------------------------------------------------------------------------
+};
+
+// </editor-fold> end GraphBuilder }}}1
+//==============================================================================
+
+//==============================================================================
+// <editor-fold desc="create_graph"> {{{1
+
+template <class ExecutionSpace, class Closure>
+Graph<ExecutionSpace> create_graph(ExecutionSpace ex, Closure&& arg_closure) {
+  // Create a shared pointer to the graph:
+  auto rv = Kokkos::Impl::GraphAccess::construct_graph(ex);
```

Comment:
Need an attorney class here so we have an implementation friend to create a graph class thingy without graph having public constructors.

Reply:
Addressed in 34490f1